tests/scale_test/notifier.py

```diff
@@ -0,0 +1,239 @@
+    utils.required_checker(ctx, file=file)
+
+    # Checking if the auth_token given is correct
+    auth_token = utils.auth_checker(ctx)
+
+    # Prepare arguments
+    bundles = utils.read_bundles(file)
+    lira_url, label, save_path = ctx.obj['lira_url'], ctx.obj['label'], ctx.obj['save_path']
+
+    if run_mode == 'async':
+        logging.info('Sending notifications asynchronously...\n')
+        async_notify(bundles, lira_url, label, auth_token, save_path)
+    elif run_mode == 'sync':
+        logging.info('Sending notifications synchronously...\n')
+        linear_notify(bundles, lira_url, label, auth_token, save_path)
+
+
+def linear_notify(bundles, lira_url, label, auth_token, save_path):
+    # Use a probe to get the current subscription_id
+    subscription_id = utils.subscription_probe(lira_url)
+    logging.info('Using subscription_id: {}'.format(subscription_id))
+
+    # Prepare the payload
+    queue = deque([])
+    for bundle in bundles:
+        queue.append(utils.prepare_notification(bundle['bundle_uuid'], bundle['bundle_version'], subscription_id, label))
+
+    # Send notifications
+    start = timeit.default_timer()
+    responses = [utils.send_notification(lira_url, auth_token, notification) for notification in tqdm(queue)]
+    stop = timeit.default_timer()
+
+    # Save metrics
+    rtt = [response.elapsed.total_seconds() for response in responses]
+    status = [response.status_code for response in responses]
+    total_time = stop-start
+    save_file = pathlib.Path(save_path) / 'notifier_metrics_{}.json'.format(time.strftime("%Y%m%d-%H%M%S"))
+    utils.dump_metrics(save_file, rtt=rtt, status=status, total_time=total_time)
+
+
+def async_notify(bundles, lira_url, label, auth_token, save_path):
+    # Use a probe to get the current subscription_id
+    subscription_id = utils.subscription_probe(lira_url)
+    logging.info('Using subscription_id: {}'.format(subscription_id))
+
+    # Prepare the payload
+    queue = deque([])
+    for bundle in bundles:
+        queue.append(utils.prepare_notification(bundle['bundle_uuid'], bundle['bundle_version'], subscription_id, label))
+
+    # Send notifications
+    partial_send_notification = functools.partial(utils.send_notification, lira_url, auth_token)
+
+    start = timeit.default_timer()
+    with concurrent.futures.ThreadPoolExecutor(max_workers=16) as executor:
+        responses = deque(executor.map(partial_send_notification, queue))
+    stop = timeit.default_timer()
+
+    # Save metrics
+    rtt = [response.elapsed.total_seconds() for response in responses]
+    status = [response.status_code for response in responses]
+    total_time = stop - start
+    save_file = pathlib.Path(save_path) / 'notifier_metrics_{}.json'.format(time.strftime("%Y%m%d-%H%M%S"))
+    utils.dump_metrics(save_file, rtt=rtt, status=status, total_time=total_time)
+
+    logging.info('Saved the metrics file to {}'.format(save_file))
+
+
+if __name__ == '__main__':
+    cli(obj={})
```

Comment:
Can you clarify why the `cli` function doesn't take in any parameters, but is getting passed an empty dict?

Reply:
Yes, this is one of the overheads of using `click` for complex CLIs, in order to pass/carry over the context from the parent command group to nested subcommands, you have to create either a class or a data structure to store the parent parameters/info. I chose `dict` here, but it could have been `list` or even custom class. I will add some docstrings near to the code. Thanks for asking!